MediaGalleryMetadata/Model/Png/ExtractMetadata.php

```diff
@@ -0,0 +1,81 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\MediaGalleryMetadata\Model\Png;
+
+use Magento\MediaGalleryMetadataApi\Api\ExtractMetadataInterface;
+use Magento\MediaGalleryMetadataApi\Api\Data\MetadataInterface;
+use Magento\MediaGalleryMetadataApi\Api\Data\MetadataInterfaceFactory;
+use Magento\MediaGalleryMetadataApi\Model\MetadataReaderInterface;
+
+/**
+ * Extract metadata from the asset by path
+ */
+class ExtractMetadata implements ExtractMetadataInterface
+{
+    /**
+     * @var FileReader
+     */
+    private $fileReader;
+
+    /**
+     * @var MetadataReaderInterface[]
+     */
+    private $readers;
+
+    /**
+     * @var MetadataInterfaceFactory
+     */
+    private $metadataFactory;
+
+    /**
+     * @param FileReader $fileReader
+     * @param MetadataInterfaceFactory $metadataFactory
+     * @param MetadataReaderInterface[] $readers
+     */
+    public function __construct(
+        FileReader $fileReader,
+        MetadataInterfaceFactory $metadataFactory,
+        array $readers
+    ) {
+        $this->readers = $readers;
+        $this->fileReader = $fileReader;
+        $this->metadataFactory = $metadataFactory;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function execute(string $path): MetadataInterface
```

Comment:
All the ExtractMetadata classes seem to be doing exactly the same thing, to avoid code duplication wouldn't be better to use virtual type or have a builder for this?

Reply:
I am still not exactly sure the code for each extract metadata implementation is going to be exactly the same, but yes that's something I was also thinking about, probably worth going for it, thanks!

Reply:
Added virtual types